File: overlay-src/overlay.swift
// Full-screen blurred reminder overlay for Badger.
//
// Usage:
//   overlay "<item name>" <giving_up_after_seconds>
//   overlay --meeting "<title>" "<joinUrl-or-empty>" <giving_up_after_seconds>
// Checklist mode prints exactly one of: done | skip | timeout, then exits.
// Meeting mode prints exactly one of: join | dismiss | timeout, then exits.
//
// This is a compiled Swift binary (not a script run via osascript) because a
// manually-driven NSApplication invoked through `osascript -l JavaScript`
// could display windows but never actually became the key/active app, so
// real mouse clicks on the buttons were silently swallowed as mere
// focus-steal attempts. A compiled process running a real `NSApp.run()`
// event loop becomes key/main/active correctly (verified directly).
import Cocoa

let args = CommandLine.arguments
let isMeeting = args.count > 1 && args[1] == "--meeting"

let itemName: String
let givingUpAfter: Double
let meetingJoinURL: String?

if isMeeting {
    itemName = args.count > 2 ? args[2] : "Meeting"
    let rawJoinURL = args.count > 3 ? args[3] : ""
    meetingJoinURL = rawJoinURL.isEmpty ? nil : rawJoinURL
    givingUpAfter = args.count > 4 ? (Double(args[4]) ?? 90) : 90
} else {
    itemName = args.count > 1 ? args[1] : "Reminder"
    meetingJoinURL = nil
    givingUpAfter = args.count > 2 ? (Double(args[2]) ?? 50) : 50
}

final class KeyableWindow: NSWindow {
    override var canBecomeKey: Bool { true }
    override var canBecomeMain: Bool { true }
}

final class OverlayController: NSObject {
    var windows: [NSWindow] = []
    var timeoutTimer: Timer?

    func build() {
        let app = NSApplication.shared
        app.setActivationPolicy(.accessory)

        let mouseLocation = NSEvent.mouseLocation
        let activeScreen = NSScreen.screens.first { NSMouseInRect(mouseLocation, $0.frame, false) }
            ?? NSScreen.main
            ?? NSScreen.screens[0]

        for screen in NSScreen.screens {
            let window = KeyableWindow(
                contentRect: screen.frame,
                styleMask: [.borderless],
                backing: .buffered,
                defer: false,
                screen: screen
            )
            window.level = .screenSaver
            window.isOpaque = false
            window.backgroundColor = .clear
            window.collectionBehavior = [.canJoinAllSpaces, .fullScreenAuxiliary, .stationary]
            window.isReleasedWhenClosed = false

            let effect = NSVisualEffectView(frame: NSRect(origin: .zero, size: screen.frame.size))
            effect.material = .fullScreenUI
            effect.blendingMode = .behindWindow
            effect.state = .active
            window.contentView = effect

            if screen === activeScreen {
                addControls(to: effect)
            }

            window.makeKeyAndOrderFront(nil)
            windows.append(window)
        }

        app.activate(ignoringOtherApps: true)

        timeoutTimer = Timer.scheduledTimer(withTimeInterval: givingUpAfter, repeats: false) { [weak self] _ in
            self?.finish(with: "timeout")
        }

        if let selfTest = ProcessInfo.processInfo.environment["BADGER_OVERLAY_SELF_TEST"] {
            scheduleSelfTest(selfTest)
        }
    }

    var doneButton: NSButton?
    var skipButton: NSButton?
    var joinButton: NSButton?
    var dismissButton: NSButton?

    func addControls(to effect: NSVisualEffectView) {
        if isMeeting {
            addMeetingControls(to: effect)
        } else {
            addChecklistControls(to: effect)
        }
    }

    func addChecklistControls(to effect: NSVisualEffectView) {
        let label = NSTextField(labelWithString: "Reminder: \(itemName)")
        label.font = NSFont.systemFont(ofSize: 40, weight: .semibold)
        label.textColor = .white
        label.alignment = .center
        label.translatesAutoresizingMaskIntoConstraints = false
        effect.addSubview(label)

        let done = NSButton(title: "Done for today", target: self, action: #selector(onDone))
        done.bezelStyle = .rounded
        done.controlSize = .large
        done.keyEquivalent = "\r"
        done.translatesAutoresizingMaskIntoConstraints = false
        effect.addSubview(done)
        doneButton = done

        let skip = NSButton(title: "Skip today", target: self, action: #selector(onSkip))
        skip.bezelStyle = .rounded
        skip.controlSize = .large
        skip.translatesAutoresizingMaskIntoConstraints = false
        effect.addSubview(skip)
        skipButton = skip

        NSLayoutConstraint.activate([
            label.centerXAnchor.constraint(equalTo: effect.centerXAnchor),
            label.centerYAnchor.constraint(equalTo: effect.centerYAnchor, constant: -80),

            done.centerYAnchor.constraint(equalTo: effect.centerYAnchor, constant: 20),
            done.trailingAnchor.constraint(equalTo: effect.centerXAnchor, constant: -10),
            done.widthAnchor.constraint(equalToConstant: 180),

            skip.centerYAnchor.constraint(equalTo: effect.centerYAnchor, constant: 20),
            skip.leadingAnchor.constraint(equalTo: effect.centerXAnchor, constant: 10),
            skip.widthAnchor.constraint(equalToConstant: 180),
        ])
    }

    func addMeetingControls(to effect: NSVisualEffectView) {
        let label = NSTextField(labelWithString: "Starting soon: \(itemName)")
        label.font = NSFont.systemFont(ofSize: 40, weight: .semibold)
        label.textColor = .white
        label.alignment = .center
        label.translatesAutoresizingMaskIntoConstraints = false
        effect.addSubview(label)

        let dismiss = NSButton(title: "Dismiss", target: self, action: #selector(onDismiss))
        dismiss.bezelStyle = .rounded
        dismiss.controlSize = .large
        dismiss.translatesAutoresizingMaskIntoConstraints = false
        effect.addSubview(dismiss)
        dismissButton = dismiss

        if meetingJoinURL != nil {
            let join = NSButton(title: "Join Meeting", target: self, action: #selector(onJoin))
            join.bezelStyle = .rounded
            join.controlSize = .large
            join.keyEquivalent = "\r"
            join.translatesAutoresizingMaskIntoConstraints = false
            effect.addSubview(join)
            joinButton = join

            NSLayoutConstraint.activate([
                label.centerXAnchor.constraint(equalTo: effect.centerXAnchor),
                label.centerYAnchor.constraint(equalTo: effect.centerYAnchor, constant: -80),

                join.centerYAnchor.constraint(equalTo: effect.centerYAnchor, constant: 20),
                join.trailingAnchor.constraint(equalTo: effect.centerXAnchor, constant: -10),
                join.widthAnchor.constraint(equalToConstant: 180),

                dismiss.centerYAnchor.constraint(equalTo: effect.centerYAnchor, constant: 20),
                dismiss.leadingAnchor.constraint(equalTo: effect.centerXAnchor, constant: 10),
                dismiss.widthAnchor.constraint(equalToConstant: 180),
            ])
        } else {
            NSLayoutConstraint.activate([
                label.centerXAnchor.constraint(equalTo: effect.centerXAnchor),
                label.centerYAnchor.constraint(equalTo: effect.centerYAnchor, constant: -80),

                dismiss.centerXAnchor.constraint(equalTo: effect.centerXAnchor),
                dismiss.centerYAnchor.constraint(equalTo: effect.centerYAnchor, constant: 20),
                dismiss.widthAnchor.constraint(equalToConstant: 180),
            ])
        }
    }

    func scheduleSelfTest(_ mode: String) {
        Timer.scheduledTimer(withTimeInterval: 1.0, repeats: false) { [weak self] _ in
            guard let self else { return }
            switch mode {
            case "done": self.doneButton?.performClick(nil)
            case "skip": self.skipButton?.performClick(nil)
            case "join": self.joinButton?.performClick(nil)
            case "dismiss": self.dismissButton?.performClick(nil)
            default: break
            }
        }
    }

    @objc func onDone() { finish(with: "done") }
    @objc func onSkip() { finish(with: "skip") }

    @objc func onJoin() {
        if let urlString = meetingJoinURL, let url = URL(string: urlString) {
            NSWorkspace.shared.open(url)
        }
        finish(with: "join")
    }

    @objc func onDismiss() { finish(with: "dismiss") }

    func finish(with value: String) {
        timeoutTimer?.invalidate()
        for window in windows { window.orderOut(nil) }
        print(value)
        exit(0)
    }
}

let controller = OverlayController()
controller.build()
NSApplication.shared.run()

File: local_nagger/overlay.js
// JXA (JavaScript for Automation) full-screen blurred reminder overlay.
// Run via: osascript -l JavaScript overlay.js "<item name>" "<giving_up_after_seconds>"
// Prints exactly one of: done | skip | timeout
//
// Implemented as JXA (not a plain Python/PyObjC NSApplication) because this
// process is what actually owns a live WindowServer connection reliably in
// this environment — a bare PyObjC NSApplication.run() invoked from a
// background-spawned Python process was observed to hang indefinitely with
// no window ever appearing, while `osascript -e 'display dialog ...'`
// (and therefore JXA, which runs inside the same osascript host process)
// reliably shows and dismisses windows.
ObjC.import("Cocoa");

function run(argv) {
  var name = argv[0] || "Reminder";
  var givingUpAfter = parseFloat(argv[1]);
  if (isNaN(givingUpAfter)) givingUpAfter = 50;

  var state = { value: "timeout", done: false };

  ObjC.registerSubclass({
    name: "LocalNaggerOverlayController",
    methods: {
      "onDone:": {
        types: ["void", ["id"]],
        implementation: function (sender) {
          state.value = "done";
          state.done = true;
        },
      },
      "onSkip:": {
        types: ["void", ["id"]],
        implementation: function (sender) {
          state.value = "skip";
          state.done = true;
        },
      },
    },
  });

  var controller = $.LocalNaggerOverlayController.alloc.init;

  var app = $.NSApplication.sharedApplication;
  app.setActivationPolicy($.NSApplicationActivationPolicyAccessory);

  var collectionBehavior =
    $.NSWindowCollectionBehaviorCanJoinAllSpaces |
    $.NSWindowCollectionBehaviorFullScreenAuxiliary |
    $.NSWindowCollectionBehaviorStationary;

  var windows = [];
  var screens = $.NSScreen.screens;
  var screenCount = screens.count;

  for (var i = 0; i < screenCount; i++) {
    var screen = screens.objectAtIndex(i);
    var frame = screen.frame;
    var w = frame.size.width;
    var h = frame.size.height;

    var window = $.NSWindow.alloc.initWithContentRectStyleMaskBackingDeferScreen(
      frame,
      $.NSWindowStyleMaskBorderless,
      $.NSBackingStoreBuffered,
      false,
      screen
    );
    window.level = $.NSScreenSaverWindowLevel;
    window.opaque = false;
    window.backgroundColor = $.NSColor.clearColor;
    window.collectionBehavior = collectionBehavior;
    window.releasedWhenClosed = false;

    var effect = $.NSVisualEffectView.alloc.initWithFrame($.NSMakeRect(0, 0, w, h));
    effect.material = $.NSVisualEffectMaterialFullScreenUI;
    effect.blendingMode = $.NSVisualEffectBlendingModeBehindWindow;
    effect.state = $.NSVisualEffectStateActive;
    window.contentView = effect;

    var label = $.NSTextField.alloc.initWithFrame($.NSMakeRect(0, h / 2 + 10, w, 70));
    label.stringValue = "Reminder: " + name;
    label.alignment = $.NSTextAlignmentCenter;
    label.font = $.NSFont.systemFontOfSize(40);
    label.bezeled = false;
    label.drawsBackground = false;
    label.editable = false;
    label.selectable = false;
    label.textColor = $.NSColor.whiteColor;
    effect.addSubview(label);

    // Buttons only need to be interactive on one screen (the user only has
    // one mouse cursor); the rest are decorative blur-only overlays so the
    // whole desktop is blocked no matter which screen is being looked at.
    if (i === 0) {
      var doneButton = $.NSButton.alloc.initWithFrame($.NSMakeRect(w / 2 - 170, h / 2 - 70, 150, 46));
      doneButton.title = "Done for today";
      doneButton.bezelStyle = $.NSBezelStyleRounded;
      doneButton.target = controller;
      doneButton.action = "onDone:";
      doneButton.keyEquivalent = "\r";
      effect.addSubview(doneButton);

      var skipButton = $.NSButton.alloc.initWithFrame($.NSMakeRect(w / 2 + 20, h / 2 - 70, 150, 46));
      skipButton.title = "Skip today";
      skipButton.bezelStyle = $.NSBezelStyleRounded;
      skipButton.target = controller;
      skipButton.action = "onSkip:";
      effect.addSubview(skipButton);
    }

    window.makeKeyAndOrderFront($());
    windows.push(window);
  }

  app.activateIgnoringOtherApps(true);

  var deadline = $.NSDate.dateWithTimeIntervalSinceNow(givingUpAfter);
  while (!state.done && deadline.timeIntervalSinceNow > 0) {
    var untilDate = $.NSDate.dateWithTimeIntervalSinceNow(0.15);
    var event = app.nextEventMatchingMaskUntilDateInModeDequeue(
      $.NSUIntegerMax,
      untilDate,
      $.NSDefaultRunLoopMode,
      true
    );
    if (!event.isNil()) app.sendEvent(event);
  }

  windows.forEach(function (w) {
    w.orderOut($());
  });

  return state.value;
}
